### PR TITLE
feat: add ip rule

### PR DESCRIPTION
**Pull request by love98ooo**
Part of: https://github.com/casbin/caswaf/issues/30

### CODE REVIEW

File: util/request.go [hunk 1]
```diff
@@ -0,0 +1,26 @@
+package util
```
Comment (nomeguy) [inline util/request.go:15]: Apache header

File: web/src/components/IpRuleTable.js [hunk 2]
```diff
@@ -53,22 +85,28 @@ class IPRuleTable extends React.Component {
...
+
   renderTable(table) {
     const columns = [
       {
         title: "Name",
         dataIndex: "name",
         key: "name",
         width: "180px",
-        render: (text, record, index) => (
+        render: (text, rule, index) => (
```
Comment (nomeguy) [inline web/src/components/IpRuleTable.js]: keep record

File: web/src/components/IpRuleTable.js [hunk 3]
```diff
@@ -79,7 +117,7 @@ class IPRuleTable extends React.Component {
         dataIndex: "operator",
         key: "operator",
         width: "180px",
-        render: (text, record, index) => (
+        render: (text, rule, index) => (
```
Comment (nomeguy) [inline web/src/components/IpRuleTable.js]: keep record

File: web/src/components/IpRuleTable.js [hunk 4]
```diff
@@ -97,19 +135,22 @@ class IPRuleTable extends React.Component {
         dataIndex: "value",
         key: "value",
         width: "100%",
-        render: (text, record, index) => (
-          <Input value={text} placeholder="Split with Space" onChange={e => {
-            this.updateField(table, index, "value", e.target.value);
-          }} onBlur={e => {
-            this.updateField(table, index, "value", e.target.value.replace(/\s+/g, " ").trim());
-          }} />
+        render: (text, rule, index) => (
```
Comment (nomeguy) [inline web/src/components/IpRuleTable.js]: keep record

File: web/src/components/IpRuleTable.js [hunk 5]
```diff
@@ -97,19 +135,22 @@ class IPRuleTable extends React.Component {
...
+            options={this.state.options[index]}
+          />
         ),
       },
       {
         title: "Action",
         key: "action",
         width: "100px",
-        render: (text, record, index) => (
+        render: (text, rule, index) => (
```
Comment (nomeguy) [inline web/src/components/IpRuleTable.js]: keep record

File: web/src/components/IpRuleTable.js [hunk 6]
```diff
@@ -130,6 +171,7 @@ class IPRuleTable extends React.Component {
           <div>
             {this.props.title}&nbsp;&nbsp;&nbsp;&nbsp;
             <Button style={{marginRight: "5px"}} type="primary" size="small" onClick={() => this.addRow(table)}>{"Add"}</Button>
+            <Button style={{marginRight: "5px"}} type="primary" size="small" onClick={() => this.restore()}>{"Restore Build-in IP Rules"}</Button>
```
Comment (nomeguy) [inline web/src/components/IpRuleTable.js]: Rename all to "Restore"

File: rule/rule_ip.go [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+)
+
+type IpRule struct{}
+
+func (r *IpRule) checkRule(expressions []*object.Expression, req *http.Request) (bool, string, error) {
+	clientIp := util.GetClientIp(req)
+	for _, expression := range expressions {
+		reason := fmt.Sprintf("expression matched: \"%s %s %s\"", clientIp, expression.Operator, expression.Value)
+		ips := strings.Split(expression.Value, " ")
+		op := expression.Operator == "is in"
```
Comment (nomeguy) [inline rule/rule_ip.go]: what's op?

File: rule/rule_ip.go [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+	"github.com/casbin/caswaf/util"
+)
+
+type IpRule struct{}
+
+func (r *IpRule) checkRule(expressions []*object.Expression, req *http.Request) (bool, string, error) {
+	clientIp := util.GetClientIp(req)
+	for _, expression := range expressions {
+		reason := fmt.Sprintf("expression matched: \"%s %s %s\"", clientIp, expression.Operator, expression.Value)
+		ips := strings.Split(expression.Value, " ")
```
Comment (nomeguy) [inline rule/rule_ip.go]: GetClientIp should only return one IP address, so no for loop

File: rule/rule_ip.go [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+		op := expression.Operator == "is in"
+		for _, ip := range ips {
+			_, ipNet, err := net.ParseCIDR(ip)
```
Comment (nomeguy) [inline rule/rule_ip.go]: This is wrong.

If contains "/", then parse as CIDR

Else if contains ":", then parse as IPv6 address

Else, parse as IPv4 address

For each parsing error, return error

File: web/src/components/IpRuleTable.js [hunk 10]
```diff
@@ -19,25 +19,57 @@ import * as Setting from "../Setting";
...
+      defaultRules: [
+        {
+          name: "loopback",
+          operator: "is in",
+          value: "127.0.0.1",
+        },
+        {
+          name: "lan cidr",
+          operator: "is in",
+          value: "0.0.0.0/8 10.0.0.0/8 100.64.0.0/10 127.0.0.0/8 169.254.0.0/16 172.16.0.0/12 192.0.0.0/24 192.0.2.0/24 192.88.99.0/24 192.168.0.0/16 198.18.0.0/15 198.51.100.0/24 203.0.113.0/24 224.0.0.0/3",
```
Comment (nomeguy) [inline web/src/components/IpRuleTable.js]: too many as examples, so two or three is enough. Make people hard to delete them all

**Review by nomeguy — CHANGES_REQUESTED** [review]

File: rule/rule_ip.go (hunk 7 above)
Comment (love98ooo) [reply rule/rule_ip.go]: operator (op)

Comment (nomeguy) [reply rule/rule_ip.go]: op should not be boolean

File: rule/rule_ip.go (hunk 8 above)
Comment (nomeguy) [inline rule/rule_ip.go]: Use comma to split

**Review by nomeguy — CHANGES_REQUESTED** [review]

File: rule/rule_ip.go [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+	for _, expression := range expressions {
+		reason := fmt.Sprintf("expression matched: \"%s %s %s\"", clientIp, expression.Operator, expression.Value)
+		ips := strings.Split(expression.Value, ",")
+		for _, ip := range ips {
+			if strings.Contains(ip, "/") {
+				_, ipNet, err := net.ParseCIDR(ip)
+				if err != nil {
+					return false, "", err
+				}
+				if ipNet.Contains(net.ParseIP(clientIp)) == (expression.Operator == "is in") {
```
Comment (nomeguy) [inline rule/rule_ip.go]: check error for net.ParseIP(clientIp)

File: rule/rule_ip.go [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+					return true, reason, nil
+				}
+			} else if strings.ContainsAny(ip, ".:") {
+				if (ip == clientIp) == (expression.Operator == "is in") {
```
Comment (nomeguy) [inline rule/rule_ip.go]: Never write this, expression 1 == expression 2

don't play tricks

File: rule/rule_ip.go [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+					return true, reason, nil
+				}
+			} else {
+				return false, "", fmt.Errorf("invalid ip: %s", ip)
```
Comment (nomeguy) [inline rule/rule_ip.go]: unknown IP or CIDR format: %s

File: web/src/components/IpRuleTable.js [hunk 14]
```diff
@@ -19,25 +19,57 @@ import * as Setting from "../Setting";
...
+      defaultRules: [
+        {
+          name: "loopback",
+          operator: "is in",
+          value: "127.0.0.1",
+        },
+        {
+          name: "lan cidr",
+          operator: "is in",
+          value: "10.0.0.0/8,127.0.0.0/8,192.0.0.0/24",
```
Comment (nomeguy) [inline web/src/components/IpRuleTable.js]: 10.0.0.0/8,192.168.0.0/16

**Review by nomeguy — CHANGES_REQUESTED** [review]

File: rule/rule_ip.go [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+				switch expression.Operator {
+				case "is in":
+					if ipNet.Contains(netIp) {
+						return true, reason, nil
+					}
+				case "is not in":
+					if !ipNet.Contains(netIp) {
+						return true, reason, nil
+					}
+				}
```
Comment (nomeguy) [inline rule/rule_ip.go:56]: Handle switch-default like:

image

File: rule/rule_ip.go [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+				switch expression.Operator {
+				case "is in":
+					if ip == clientIp {
+						return true, reason, nil
+					}
+				case "is not in":
+					if ip != clientIp {
+						return true, reason, nil
+					}
+				}
```
Comment (nomeguy) [inline rule/rule_ip.go:69]: Handle switch-default like:

image

**Review by nomeguy — CHANGES_REQUESTED** [review]

**Review by nomeguy — APPROVED** [review]

**Pull request merged**